internal/job_runner.go

```diff
@@ -0,0 +1,270 @@
+	meta := tvm.Metadata
+	if meta == nil || meta.CustomProperties == nil {
+		return ""
+	}
+	return meta.CustomProperties["kubernetes_pod_name"]
+}
+
+type newDataMsg struct {
+	msg        *messages.DataMessage
+	handle     string
+	resolution time.Duration
+}
+
+type dataRequest struct {
+	program string
+	respCh  chan dataResponse
+}
+
+type dataResponse struct {
+	snapshot *MetricSnapshot
+	err      error
+}
+
+type startedMsg struct {
+	comp    *signalflow.Computation
+	handle  string
+	program string
+}
+
+// SignalFlowJobRunner manages SignalFlow jobs.
+type SignalFlowJobRunner struct {
+	ctx context.Context
+
+	client *signalflow.Client
+
+	jobsByProgram           map[string]*signalflow.Computation
+	jobsByHandle            map[string]*signalflow.Computation
+	metricSnapshotsByHandle map[string]*MetricSnapshot
+
+	startedCh     chan startedMsg
+	stoppedCh     chan string
+	dataCh        chan newDataMsg
+	dataRequestCh chan dataRequest
+
+	TotalJobsStarted int64
+	TotalJobsStopped int64
+	TotalJobsErrored int64
+}
+
+func NewSignalFlowJobRunner(client *signalflow.Client) *SignalFlowJobRunner {
+	return &SignalFlowJobRunner{
+		client:                  client,
+		jobsByProgram:           make(map[string]*signalflow.Computation),
+		jobsByHandle:            make(map[string]*signalflow.Computation),
+		metricSnapshotsByHandle: make(map[string]*MetricSnapshot),
+		startedCh:               make(chan startedMsg),
+		stoppedCh:               make(chan string),
+		dataCh:                  make(chan newDataMsg),
+		dataRequestCh:           make(chan dataRequest),
+	}
+}
+
+func (jr *SignalFlowJobRunner) ReplaceOrStartJob(program string) error {
+	comp, err := jr.client.Execute(&signalflow.ExecuteRequest{
+		Program: program,
+	})
+	if err != nil {
+		atomic.AddInt64(&jr.TotalJobsErrored, 1)
+		return err
+	}
+
+	comp.MetadataTimeout = 10 * time.Second
+	// Wait for the handle to come through before sending the message to block
+	// the loop less.
+	handle := comp.Handle()
+	if handle == "" {
+		return fmt.Errorf("Could not get job handle within timeout: %v", comp.Err())
+	}
+
+	jr.startedCh <- startedMsg{
+		comp:    comp,
+		handle:  handle,
+		program: program,
+	}
+
+	klog.Infof("Started SignalFlow compuation: %s (%s)", program, handle)
+	atomic.AddInt64(&jr.TotalJobsStarted, 1)
+
+	go jr.watchJob(comp, program)
+
+	return nil
+}
+
+func (jr *SignalFlowJobRunner) StopJob(program string) error {
+	jr.stoppedCh <- program
+	return nil
+}
+
+// Run does everything in a single loop to prevent heavy churn on a mutex
+// when there are a lot of jobs running and updating values, and also avoids
+// the possiblity of data races.  This trades off mutexes for channels.
+func (jr *SignalFlowJobRunner) Run(ctx context.Context) {
+	jr.ctx = ctx
+
+	for {
+		select {
+
+		case <-jr.ctx.Done():
+			klog.Infof("SignalFlow job runner is stopping")
+			return
+
+		case startedMsg := <-jr.startedCh:
+			jr.jobsByProgram[startedMsg.program] = startedMsg.comp
```

Comment:
These mutations of the map are safe because it's effectively serializing via the channel, right?

Reply:
Yes, the map is only accessed in this loop by a single goroutine.